Add support for boost::variant in C++11 mode

dean0x7d opened the pull request:
Resolves #988. 

It also turns out that we can only support `boost::variant` from Boost 1.56 or newer since that's when it started using C++11 variadic templates. In older versions `variant` has a fixed number of template arguments with the last few being placeholders. We'd need to add workarounds for this, but it's not worth it, so I've just noted the requirement in the docs.

The non-docker trusty builds on Travis only have access to Boost 1.54, which unfortunately means CI doesn't cover C++11 mode. Any ideas how to get a newer boost on Travis without docker?

jagerman commented:
Try adding `boost-latest` to `addons->apt->sources`, and add `libboost-dev` to `addons->apt->packages`.

jagerman commented:
Scratch that; it looks like the `boost-latest` ppa isn't maintained (the latest version appears to be 1.55).

jagerman commented:
I suppose you could turn on `sudo: required` for one of the builds, then add a:

```
sudo apt-get -qy remove libboost1.54-dev &&
wget http://us.archive.ubuntu.com/ubuntu/pool/main/b/boost1.58/libboost1.58-dev_1.58.0+dfsg-5ubuntu3_amd64.deb && \
sudo dpkg -i libboost1.58-dev_1.58.0+dfsg-5ubuntu3_amd64.deb
```

If you want more recent than 1.58, the Debian `stretch` deb should also work (and like the `xenial` URL above should stay valid for a long time): `http://ftp.us.debian.org/debian/pool/main/b/boost1.62/libboost1.62-dev_1.62.0+dfsg-4_amd64.deb`.

dean0x7d commented:
It turns out that `boost::variant` can't use variadic templates on MSVC 2015 so the variant caster can't support it. I've added an additional note about MSVC 2017 being required.

Regarding the C++11 tests, I've added a `static_assert` to make sure that `::result_type` is present, which should be enough to cover that. Probably best not to complicate the Travis config for this. The main thing is that the variant caster is being tested with at least one third-party `variant` (with the C++14 docker builds).
